pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManagerForConsistentDeletes.java

```diff
@@ -0,0 +1,391 @@
+              } else {
+                return new RecordLocation(currentSegment, currentDocId, currentComparisonValue,
+                    RecordLocation.incrementSegmentCount(currentDistinctSegmentCount));
+              }
+            } else {
+              // New primary key
+              addDocId(segment, validDocIds, queryableDocIds, newDocId, recordInfo);
+              return new RecordLocation(segment, newDocId, newComparisonValue, 1);
+            }
+          });
+    }
+    int numKeys = numKeysInWrongSegment.get();
+    if (numKeys > 0) {
+      _logger.warn("Found {} primary keys in the wrong segment when adding segment: {}", numKeys, segmentName);
+      _serverMetrics.addMeteredTableValue(_tableNameWithType, ServerMeter.UPSERT_KEYS_IN_WRONG_SEGMENT, numKeys);
+    }
+  }
+
+  @Override
+  protected void addSegmentWithoutUpsert(ImmutableSegmentImpl segment, ThreadSafeMutableRoaringBitmap validDocIds,
+      @Nullable ThreadSafeMutableRoaringBitmap queryableDocIds, Iterator<RecordInfo> recordInfoIterator) {
+    throw new UnsupportedOperationException("Consistent-deletion does not support preloading of segments.");
+  }
+
+  @Override
+  protected void doRemoveSegment(IndexSegment segment) {
+    String segmentName = segment.getSegmentName();
+    _logger.info("Removing {} segment: {}, current primary key count: {}",
+        segment instanceof ImmutableSegment ? "immutable" : "mutable", segmentName, getNumPrimaryKeys());
+    long startTimeMs = System.currentTimeMillis();
+
+    try (
+        UpsertUtils.PrimaryKeyReader primaryKeyReader = new UpsertUtils.PrimaryKeyReader(segment, _primaryKeyColumns)) {
+      removeSegment(segment,
+          UpsertUtils.getPrimaryKeyIterator(primaryKeyReader, segment.getSegmentMetadata().getTotalDocs()));
+    } catch (Exception e) {
+      throw new RuntimeException(
+          String.format("Caught exception while removing segment: %s, table: %s", segment.getSegmentName(),
+              _tableNameWithType), e);
+    }
+
+    // Update metrics
+    long numPrimaryKeys = getNumPrimaryKeys();
+    updatePrimaryKeyGauge(numPrimaryKeys);
+    _logger.info("Finished removing segment: {} in {}ms, current primary key count: {}", segmentName,
+        System.currentTimeMillis() - startTimeMs, numPrimaryKeys);
+  }
+
+  @Override
+  protected void removeSegment(IndexSegment segment, Iterator<PrimaryKey> primaryKeyIterator) {
+    Set<Object> deDuplicatedPrimaryKeys = new HashSet<>();
+    while (primaryKeyIterator.hasNext()) {
+      PrimaryKey primaryKey = primaryKeyIterator.next();
+      _primaryKeyToRecordLocationMap.computeIfPresent(HashUtils.hashPrimaryKey(primaryKey, _hashFunction),
+          (pk, recordLocation) -> {
+            if (deDuplicatedPrimaryKeys.contains(pk)) {
+              return recordLocation;
+            }
+            deDuplicatedPrimaryKeys.add(pk);
+            if (recordLocation.getSegment() == segment) {
+              return null;
+            }
```

Comment:
(minor) Consider moving this check before unique key check to reduce key set size